src/backends/perf_pt/decode.c

```diff
@@ -35,13 +35,317 @@
+
+    // Instantiate a decoder.
+    decoder = pt_blk_alloc_decoder(&config);
+    if (decoder == NULL) {
+        failing = true;
+        goto clean;
+    }
+
+    // Sync the decoder.
+    *decoder_status = pt_blk_sync_forward(decoder);
+    if (*decoder_status == -pte_eos) {
+        // There were no blocks in the stream. The user will find out on next
+        // call to perf_pt_next_block().
+        goto clean;
+    } else if (*decoder_status < 0) {
+        failing = true;
+        goto clean;
+    }
+
+    // Build and load a memory image from which to recover control flow.
+    struct pt_image *image = pt_image_alloc(NULL);
+    if (image == NULL) {
+        failing = true;
+        goto clean;
+    }
+
+    struct load_self_image_args load_args = {image, vdso_fd, vdso_filename};
+    if (!load_self_image(&load_args)) {
+        failing = true;
+        goto clean;
+    }
+
+    if (pt_blk_set_image(decoder, image) < 0) {
+        failing = true;
+        goto clean;
+    }
+
+clean:
+    if (failing) {
+        pt_blk_free_decoder(decoder);
+        return NULL;
+    }
+    return decoder;
+}
+
+/*
+ * Updates `*addr' with the start address of the next block in the instruction
+ * stream.
+ *
+ * A start address of 0 indicates that the end of the instruction stream has
+ * been reached.
+ *
+ * `*decoder_status` will be updated with the new decoder status after the operation.
+ *
+ * Returns true on success or false otherwise. Upon failure, the start address
+ * is undefined.
+ */
+bool
+perf_pt_next_block(struct pt_block_decoder *decoder, int *decoder_status,
+                   uint64_t *addr) {
+    // If there are events pending, drain them first.
+    *decoder_status = drain_events(decoder, *decoder_status);
+    if (*decoder_status < 0) {
+        // Error.
+        return false;
+    } else if (*decoder_status & pts_eos) {
+        // End of stream.
+        *addr = 0;
+        return true;
+    } else if (*decoder_status > 0) {
+        // An unhandled case.
+        return false;
+    }
+
+    // Now fetch the block information.
+    struct pt_block block;
+    *decoder_status = pt_blk_next(decoder, &block, sizeof(block));
+    // Other +ve decoder status codes can arise here. We ignore them for now,
+    // and let them be detected by drain_events() above when we are next
+    // called.
+    if (*decoder_status < 0) {
+        if (*decoder_status == -pte_eos) {
+            // End of stream is flagged as an error in the case of pt_blk_next().
+            *addr = 0;
+            return true;
+        } else {
+            // A real error.
+            return false;
+        }
+    }
+
+    if (block.truncated == 1) {
+        // XXX A truncated block. This occurs when a block straddles a section
+        // boundary. In this case we may need some extra logic, but this should
+        // be a rare case.
+        return false;
+    }
+
+    if (block.ninsn == 0) {
+        // A block with zero instructions. Shouldn't happen.
```

Comment:
If it shouldn't happen, this should be an `assert`?

Reply:
It could be yeah. Do you think that's better than returning an error?

Reply:
If it's not supposed to happen, then I think so.